runtime/binding-kafka/src/main/java/io/aklivity/zilla/runtime/binding/kafka/internal/validator/AvroValidator.java

```diff
@@ -73,25 +79,77 @@ public boolean validate(
         data.getBytes(0, payloadBytes);
         ByteBuffer byteBuf = ByteBuffer.wrap(payloadBytes);
 
-        tryValidate:
+        if (byteBuf.get() != MAGIC_BYTE)
+        {
+            System.out.println("Unknown magic byte!");
+            return false;
+        }
+
+        int schemaId = byteBuf.getInt();
+        int valLength = length - 1 - 4;
+        byte[] valBytes = new byte[valLength];
+        data.getBytes(length - valLength, valBytes);
+
+        ParsedSchema parsedSchema = handler.resolve(schemaId);
+
+        if (parsedSchema != null && validate(valBytes, parsedSchema.schema))
+        {
+            status = true;
+        }
+        return status;
+    }
+
+    @Override
+    public boolean write(
+        DirectBuffer data,
+        int index,
+        int length,
+        boolean isKey)
+    {
+        boolean status = false;
+        int schemaId = catalog.schemaId;
+
+        byte[] payloadBytes = new byte[length];
+        data.getBytes(0, payloadBytes);
+
         try
         {
-            if (byteBuf.get() != MAGIC_BYTE)
+            ParsedSchema parsedSchema = null;
+            if (schemaId > 0)
             {
-                System.out.println("Unknown magic byte!");
-                break tryValidate;
+                parsedSchema = handler.resolve(schemaId);
+
+            }
+            else if (catalog.strategy.equals("topic"))
+            {
+                parsedSchema = handler.resolve(isKey ? keySubject : valueSubject, catalog.version);
```

Comment:
We discussed having `handler.resolve(subject, version)` return the `schemaId` so that `handler.resolve(schemaId)` would be used in all cases.

If we do that, then it seems we don't need `ParsedSchema`, agree?